### PR TITLE
Pushdown DATE, TIMESTAMP, and TIMESTAMP_TZ_MILLIS predicates in mongodb connector

### PR DESCRIPTION
**Pull request by posulliv**
Resolves: #5658

### CODE REVIEW

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 1]
```diff
@@ -407,6 +409,14 @@ else if (singleValues.size() > 1) {
             return Optional.of(prestoNativeValue);
         }
 
+        if (type == DATE) {
+            return Optional.of(new Date(TimeUnit.DAYS.toMillis((Long) prestoNativeValue)));
+        }
+
+        if (type == TIMESTAMP_MILLIS) {
+            return Optional.of(new Date(TimeUnit.MILLISECONDS.convert((Long) prestoNativeValue, TimeUnit.MICROSECONDS)));
+        }
```
Comment (ebyhr) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:418]: Can we add support for `timestamp with time zone` as well?

Comment (posulliv) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:420]: @ebyhr yep, updated to add that.

**Comment by zifeo** [comment]
@posulliv I worked on the same feature in parallel, you may want to have a [look](https://github.com/prestosql/presto/compare/master...zifeo:timestamp-mongo?expand=1).

File: presto-mongodb/src/test/java/io/prestosql/plugin/mongodb/TestMongoIntegrationSmokeTest.java [hunk 2]
```diff
@@ -357,6 +359,41 @@ public void testNullPredicates()
...
+                .matches("VALUES (BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25)))")
+                .isNotFullyPushedDown(FilterNode.class);
+
+        // bigint range, with decimal to bigint simplification
+        assertThat(query("SELECT regionkey, nationkey, name FROM nation WHERE nationkey BETWEEN 18.5 AND 19.5"))
+                .matches("VALUES (BIGINT '3', BIGINT '19', CAST('ROMANIA' AS varchar(25)))")
+                .isNotFullyPushedDown(FilterNode.class);
+
+        // date equality
+        assertThat(query("SELECT orderkey FROM orders WHERE orderdate = DATE '1992-09-29'"))
```
Comment (zifeo) [inline presto-mongodb/src/test/java/io/prestosql/plugin/mongodb/TestMongoIntegrationSmokeTest.java:392]: Can you add a case for the timestamp and timestamp with timezone?

**Comment by posulliv** [comment]
@zifeo thanks for the heads up! I can revert my changes to account for timestamp type if you want and you can open a PR with your changes?

**Comment by zifeo** [comment]
@posulliv your implementation looks better/lighter in my opinion, not really used to those kind of changes. You can pick changes on mine if you need some elements.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 3]
```diff
@@ -407,6 +411,18 @@ else if (singleValues.size() > 1) {
...
+
+        if (type == TIMESTAMP_TZ_MILLIS) {
+            return Optional.of(new Date(unpackMillisUtc((Long) prestoNativeValue)));
```
Comment (findepi) [inline presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:423]:
```suggestion
            return Optional.of(new Date(unpackMillisUtc((long) prestoNativeValue)));
```

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 4]
```diff
@@ -407,6 +411,18 @@ else if (singleValues.size() > 1) {
             return Optional.of(prestoNativeValue);
         }
 
+        if (type == DATE) {
+            return Optional.of(new Date(TimeUnit.DAYS.toMillis((Long) prestoNativeValue)));
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:415]: Similarly, i think the conversion here
https://github.com/prestosql/presto/blob/22717c61419c851b1c3814b60a3453b22d4a45ef/presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoPageSink.java#L167
may be wrong.

File: presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java [hunk 5]
```diff
@@ -407,6 +411,18 @@ else if (singleValues.size() > 1) {
...
+        }
+
+        if (type == TIMESTAMP_MILLIS) {
+            return Optional.of(new Date(TimeUnit.MILLISECONDS.convert((Long) prestoNativeValue, TimeUnit.MICROSECONDS)));
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoSession.java:419]: Similarly, i think the conversion here
https://github.com/prestosql/presto/blob/22717c61419c851b1c3814b60a3453b22d4a45ef/presto-mongodb/src/main/java/io/prestosql/plugin/mongodb/MongoPageSink.java#L174-L175
may be wrong.

**Comment by kurtqq** [comment]
any progress with this PR?

**Comment by Deninc** [comment]
Is this not reviewed anymore?

**Comment by cansevilmis** [comment]
Any update on this? 
thanks

**Comment by harlequin** [comment]
I have had made a review about the PR.

## Comment to change `Long` to `long`

Would make sense due to the fact the `long` is primitive and have to have an value. 
But not really needed to accept this change.

## Comments about the `Date`
 
@findepi : 
Yes you are absolute right, all Date Time operations are working well in case the the JVM is running in UTC.
But for this change we can't introduce the LocalDate and LocalDateTime (which is supported since MongoDb Driver version 3.7)
without changing a lot inside the complete connector.


If you taken both investigations into consideration, I would propose to merge it.
Because :-
1. the scope of this PR is to pushdown date fields (which is working fine, I have performed some tests on my machine)
2. The whole connector is only working correct with JVM UTC

I would also propose to write a new Issue which will focus on the rework of all `java.util.Date` imports and to work with the newer `LocalDate`

**Review by ebyhr — COMMENTED** [review]
Could you please update `TestMongoTypeMapping`?

**Comment by ebyhr** [comment]
Superseded by #14413

**Pull request closed without merging**